fixes #2260 Allow ERB in ENC global / class parameters

GregSutcliffe opened the pull request:
Adds code to parse the global & class parameters, and adds a setting (default true)
 to disable this if needed. Also has some tests.

GregSutcliffe committed:
fixes #2260 Allow ERB in ENC global / class parameters

Adds code to parse the global & class parameters, and adds a setting (default true)
 to disable this if needed. Also has some tests.

ohadlevy commented:
I would render write a separate class which accepts a host and a string and evaluate it.
imho, it would be easier to follow, and maybe we could simplify some other places that does the same (e.g. ptable disk layout etc).

GregSutcliffe commented:
@ohadlevy fair point, i'll give that a look.

ohadlevy commented on lib/foreman/parse_erb_for_host.rb:
```diff
@@ -0,0 +1,13 @@
+# Tiny class to parse ERB in parameter values
+#
```
this is not a class ;)

ohadlevy commented on lib/foreman/parse_erb_for_host.rb:
```diff
@@ -0,0 +1,13 @@
+#
+module Foreman
+  module ParseErbForHost
+    include Foreman::Renderer
```
I had in mind a real class, which in the renderer stuff is in it, something like

``` ruby
SafeRender.new(:string => string)
```

GregSutcliffe commented:
@ohadlevy ok, moved to a real class. I couldn't get

```
SafeRender.new(:string => string)
```

to actually return a string, and we need to pass in the host object anyway, so I've gone with

```
SafeRender.new(<context>).parse(string)
```

Hope that's ok.

ohadlevy commented:
this looks good, is it much work to get the  rendere code out of there and use the same code for unattended?

GregSutcliffe commented:
Probably not, I'll take a shot at that now.

GregSutcliffe commented:
@ohadlevy spent an hour or two on this, and got stuck in a mess of undefined instance variables. It's probably something easy to fix, but right now I want to get back to the discovery stuff. Can we merge this as is and add a refactor task to the backlog?

ohadlevy commented:
when you have some free time, can you rebase please?

GregSutcliffe commented:
@ohadlevy rebased

GregSutcliffe commented:
@ohadlevy rebased again ;)

domcleal commented:
@GregSutcliffe tests have failed on MySQL (I'd ignore the postgres failure, it looks random).  I suspect you just need to quote "true" in the settings fixtures file rather than bareword.

GregSutcliffe commented:
@domcleal thanks, i've fixed that. @ohadlevy I've added handling for recursing through Array- and Hash-valued data (and tests for that) - this is due to the conversation at https://groups.google.com/d/msg/foreman-users/ntubzQSzRio/QnCYZpAhCiIJ - can you re-review?

GregSutcliffe commented:
thanks @witlessbird for the fixes, @ohadlevy review?

domcleal commented on app/models/setting/puppet.rb:
```diff
@@ -21,6 +21,7 @@ def self.load_defaults
         self.set('Default_variables_Lookup_Path', N_("The Default path in which Foreman resolves host specific variables"), ["fqdn", "hostgroup", "os", "domain"]),
         self.set('Enable_Smart_Variables_in_ENC', N_("Should the smart variables be exposed via the ENC yaml output?"), true),
         self.set('Parametrized_Classes_in_ENC', N_("Should Foreman use the new format (2.6.5+) to answer Puppet in its ENC yaml output?"), param_enc),
+        self.set('interpolate_erb_in_parameters', "Should Foreman parse ERB to return dynamic parameters?", true),
```
missing `N_(..)`

domcleal commented on app/helpers/unattended_helper.rb:
```diff
@@ -13,4 +16,34 @@ def root_pass
+      begin
+        return default_safe_render(template.template)
+      rescue Exception => exc
+        raise "The snippet '#{name}' threw an error: #{exc}"
```
ideally these would now be `raise Foreman::Exception.new("The snippet '%{name}' threw an error: %{e}", { :name => name, :e => exc })`

domcleal commented on app/models/config_template.rb:
```diff
@@ -105,7 +104,7 @@ def self.build_pxe_default(renderer)
     if error_msg.empty?
       begin
         @profiles = pxe_default_combos
-        menu = renderer.render_safe(default_template.template, [:default_template_url], {:profiles => @profiles})
```
tidyup possible here, the renderer argument can be removed and from the call locations in the controllers

GregSutcliffe replied:
Not sure I follow, can you elaborate?

domcleal replied:
this method's accepting a renderer argument, which it no longer uses as it instantiates one instead

GregSutcliffe replied:
I'm still not sure I follow - we're passing in one hash and one array argument, to override the defaults in default_safe_render.

GregSutcliffe replied:
I assume you're commenting on the line below, since this one is the deleted line ;)

domcleal replied:
no, I'm commenting on the method we're in, build_pxe_default

GregSutcliffe replied:
Ah gotcha, will fix.

GregSutcliffe commented:
@domcleal fixed the i18n and exceptions, need a little guidance on the last one.

ohadlevy commented on app/helpers/unattended_helper.rb:
```diff
@@ -13,4 +16,34 @@ def root_pass
+    url_for :only_path => false, :controller => "/unattended", :action => action,
+      :host      => (Setting[:foreman_url] unless Setting[:foreman_url].blank?),
+      :protocol  => 'http',
+      :token     => (@host.token.value unless @host.token.nil?)
```
what if @host is not defined?

dmitri-d replied:
Same for :token, however I don't think this is as severe. Is there a default value for Setting[:foreman_url]? Another solution is to throw an exception if :foreman_url is not defined - I don't think trying to guess it is going to work...

ohadlevy replied:
this is actually important as the render can operate outside the scope of a host, e.g. the default pxe template

dmitri-d replied:
in case of default pxe template foreman_url is not going to be used though. Perhaps runtime exception during template rendering is not such a bad idea (provided the error message is sufficiently informative)?

GregSutcliffe replied:
For the token, I think it's sufficient to rescue to nil, as if we don't have `@host` we _definitely_ don't want a token either.

dmitri-d replied:
@GregSutcliffe gotcha re: token. Is there any case when we'd use foreman_url without a host though?

GregSutcliffe replied:
At the moment the foreman codebase has two separate foreman_url functions defined (one is a helper, one is in lib/) - iirc, one of them can be called without `@host`. The most obvious example is the PXE Default File template, which is written without knowing `@host` - obviously, because it's a default.

So maybe something like `@host.token.value rescue nil` ?

dmitri-d replied:
Apologies, I didn't phrase the question correctly: Is there a reason to use foreman_url when :host is not available? Would anyone need paths only?  If so, ":only_path => false" should change depending on the presence of :host...

GregSutcliffe replied:
ah, you're talking about `:host` not `@host` :)

So the history here is that templates should _not_ be dependant on the browser of the person editing the template (or building the host). Just because I reach Foreman on "http://myforwarder:5000" doesn't mean that the machine I'm building does the same.

So, we use `:only_path => false` to strip out the users route to foreman, and replace it with the value from Settings, to ensure the templates are always consistent.

GregSutcliffe replied:
I guess I haven't really answered your question though. `:host` is based on Settings, and iirc that's not one of the Settings which can be set to blank. So `:host` should always exist. It might be invalid, but that's the user's problem ;)

GregSutcliffe commented:
For the record, i've removed the refactoring commits from this as it's too risky this close to 1.2. Those commits are now at https://github.com/GregSutcliffe/foreman/tree/refactor_renderer. @domcleal can you review just the ERB part now?

GregSutcliffe commented:
Rebased following the Great Lib Cleanup of 2013

abohne commented:
Any chance of this getting merged soon?  Would be nice to not manually apply this patch anymore :smile:

ohadlevy commented on app/models/safe_render.rb:
```diff
@@ -0,0 +1,41 @@
+#   @allowed_vars = { :host => @host }
+# so that <%= @host.name %> has the right @host variable
+#
+class SafeRender
```
this should be relocated to app/services

ohadlevy commented on app/models/host/managed.rb:
```diff
@@ -299,6 +299,10 @@ def info
 
+    # Parse ERB values contained in the parameters
+    s=SafeRender.new(:variables => { :host => self} )
+    param.each { |k,v| param[k] = s.parse(v) }
```
any reason to do it a param at a time, instead of passing a hash to the render class?

I think it would be nicer to hide that from the host.

GregSutcliffe replied:
I guess we could pass the whole hash to the host, let me try that.

ohadlevy commented:
@GregSutcliffe planning to merge this for 1.3 if possible

GregSutcliffe commented:
@ohadlevy rebased, both comments addressed

ohadlevy commented on app/services/classification/base.rb:
```diff
@@ -72,9 +73,9 @@ def values_hash options={}
+        @saferender.parse(values[key.id][key.to_s][:value])
       else
-        key.default_value
+        @saferender.parse(key.default_value)
```
cant you do it after the value is defined? no need to write it twice.

ohadlevy replied:
shouldnt we do it only if the setting is enabled?

GregSutcliffe replied:
Saferender checks the settings and returns the original string immediately if the Setting is false

ohadlevy commented on app/services/safe_render.rb:
```diff
@@ -0,0 +1,41 @@
+      return object
+    end
+  end
+
```
private?

GregSutcliffe replied:
[done] you mean after `parse`? Parse is definitely public, I guess parse_string could be private

ohadlevy commented on app/services/safe_render.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def parse_string string
+    return unless string.is_a? String
```
raise?

GregSutcliffe replied:
Yep, will do

GregSutcliffe commented:
@ohadlevy refactored slightly

ohadlevy commented:
refactored and merged as 84ae960, for 1.3 thanks @GregSutcliffe
